File: object detection/birds eye viewer/cpp/src/BatchSystemHandler.cpp
#include "BatchSystemHandler.hpp"

BatchSystemHandler::BatchSystemHandler(int data_retention_time) {
  batch_data_retention = data_retention_time;
}

BatchSystemHandler::~BatchSystemHandler()
{
   clear();
}

void BatchSystemHandler::clear() {
    objects_tracked_queue.clear();
    camWorldPoseMap_ms.clear();
    camLocalPoseMap_ms.clear();

    while(!imageMap_ms.empty()) {
        std::map<unsigned long long,sl::Mat>::iterator it  = imageMap_ms.begin();
        it->second.free();
        imageMap_ms.erase(it);
    }

    while(!depthMap_ms.empty()) {
        std::map<unsigned long long,sl::Mat>::iterator it  = depthMap_ms.begin();
        it->second.free(sl::MEM::GPU);
        depthMap_ms.erase(it);
    }
}

///
/// \brief push: push data in the FIFO system
/// \param local_pose_ : current pose of the camera in Camera reference frame
/// \param world_pose_ : current pose of the camera in World reference frame
/// \param image_ : as sl::Mat (on CPU memory) to be stored
/// \param pc_ : point cloud as sl::Mat (on GPU memory) to be stored
/// \param batch_ : batch_ from ZED SDK batching system
///
void BatchSystemHandler::push(sl::Pose local_pose_, sl::Pose world_pose_, sl::Mat image_, sl::Mat pc_, std::vector<sl::ObjectsBatch> &batch_) {
    ingestWorldPoseInMap(world_pose_);
    ingestLocalPoseInMap(local_pose_);
#if WITH_IMAGE_RETENTION
    ingestImageInMap(world_pose_.timestamp,image_);
    ingestDepthInMap(world_pose_.timestamp,pc_);
#endif
    ingestInObjectsQueue(batch_);
}

///
/// \brief pop : pop the data from the FIFO system
/// \param local_pose_ : pose at the sl::Objects timestamp in camera reference frame
/// \param world_pose_ : pose at the sl::Objects timestamp in world reference frame
/// \param image_ : as sl::Mat (on CPU memory) at the sl::Objects timestamp
/// \param depth_ : point cloud as sl::Mat (on GPU memory) at the sl::Objects timestamp
/// \param objects_ : sl::Objects in the past.
///
void BatchSystemHandler::pop(sl::Pose& local_pose_,sl::Pose& world_pose_,sl::Mat& image_, sl::Mat& depth_,sl::Objects& objects_) {
    memset(&objects_,0,sizeof(sl::Objects));
    memset(&local_pose_,0,sizeof(sl::Pose));
    memset(&world_pose_,0,sizeof(sl::Pose));

    if (objects_tracked_queue.size()>0) {
        sl::Objects tracked_merged_obj = objects_tracked_queue.front();
        if (init_queue_ts.data_ns==0ULL)
            init_queue_ts = tracked_merged_obj.timestamp;

        unsigned long long targetTS_ms = tracked_merged_obj.timestamp.getMilliseconds();

        local_pose_ =  findClosestLocalPoseFromTS(targetTS_ms);
        world_pose_ =  findClosestWorldPoseFromTS(targetTS_ms);

        #if WITH_IMAGE_RETENTION
        sl::Mat tmp_image=findClosestImageFromTS(targetTS_ms);
        tmp_image.copyTo(image_);
        tmp_image.free();
        imageMap_ms[targetTS_ms].free();
        imageMap_ms.erase(targetTS_ms);

        sl::Mat tmp_depth=findClosestDepthFromTS(targetTS_ms);
        tmp_depth.copyTo(depth_,sl::COPY_TYPE::GPU_GPU);
        tmp_depth.free();
        depthMap_ms[targetTS_ms].free(sl::MEM::GPU);
        depthMap_ms.erase(targetTS_ms);
        #endif
        objects_ = tracked_merged_obj;
        objects_tracked_queue.pop_front();
    }
}

///
/// \brief push: push data in the FIFO system. Overloaded fct for objects data only
/// \param batch_ : batch_ from ZED SDK batching system
///
void BatchSystemHandler::push(std::vector<sl::ObjectsBatch> &batch_) {
    ingestInObjectsQueue(batch_);
}

///
/// \brief pop : pop the data from the FIFO system. Overloaded fct for objects data only
/// \param objects_ : sl::Objects in the past.
///
void BatchSystemHandler::pop(sl::Objects& objects_) {
    memset(&objects_,0,sizeof(sl::Objects));
    if (objects_tracked_queue.size()) {
        sl::Objects tracked_merged_obj = objects_tracked_queue.front();
        objects_ = tracked_merged_obj;
        objects_tracked_queue.pop_front();
    }
}

///
/// \brief ingestWorldPoseInMap
/// \param pose : sl::Pose of the camera in world reference frame
///
void BatchSystemHandler::ingestWorldPoseInMap(sl::Pose pose) {
    std::map<unsigned long long,sl::Pose>::iterator it = camWorldPoseMap_ms.begin();
    sl::Timestamp ts = pose.timestamp;
    if (init_app_ts.data_ns==0ULL)
        init_app_ts =  ts;

    for(auto it = camWorldPoseMap_ms.begin(); it != camWorldPoseMap_ms.end(); ) {
        if(it->first<ts.getMilliseconds() - (unsigned long long)batch_data_retention*1000)
            it = camWorldPoseMap_ms.erase(it);
        else
            ++it;
    }

    camWorldPoseMap_ms[ts.getMilliseconds()]=pose;
}

///
/// \brief ingestLocalPoseInMap
/// \param pose : sl::Pose of the camera in camera reference frame
///
void BatchSystemHandler::ingestLocalPoseInMap(sl::Pose pose) {
    std::map<unsigned long long,sl::Pose>::iterator it = camLocalPoseMap_ms.begin();
    sl::Timestamp ts = pose.timestamp;
    if (init_app_ts.data_ns==0ULL)
        init_app_ts =  ts;

    for(auto it = camLocalPoseMap_ms.begin(); it != camLocalPoseMap_ms.end(); ) {
        if(it->first<ts.getMilliseconds() - (unsigned long long)batch_data_retention*1000)
            it = camLocalPoseMap_ms.erase(it);
        else
            ++it;
    }

    camLocalPoseMap_ms[ts.getMilliseconds()]=pose;
}

void BatchSystemHandler::ingestImageInMap(sl::Timestamp ts, sl::Mat &image) {
    imageMap_ms[ts.getMilliseconds()].clone(image);
    for(auto it = imageMap_ms.begin(); it != imageMap_ms.end(); ) {
        if(it->first<ts.getMilliseconds() - (unsigned long long)batch_data_retention*1000*2)
        {
            it->second.free();
            it = imageMap_ms.erase(it);
        }
        else
            ++it;
    }
}

void BatchSystemHandler::ingestDepthInMap(sl::Timestamp ts, sl::Mat &depth) {
    depthMap_ms[ts.getMilliseconds()].clone(depth);
    for(auto it = depthMap_ms.begin(); it != depthMap_ms.end(); ) {
        if(it->first<ts.getMilliseconds() - (unsigned long long)batch_data_retention*1000*2) {
            it->second.free(sl::MEM::GPU);
            it = depthMap_ms.erase(it);
        }
        else
            ++it;
    }
}

///
/// \brief findClosestPoseFromTS : find the sl::Pose that matched the given timestamp
/// \param timestamp in milliseconds. ( at least in the same unit than camPoseMap_ms)
/// \return sl::Pose found.
///
sl::Pose BatchSystemHandler::findClosestWorldPoseFromTS(unsigned long long timestamp) {
    sl::Pose pose = sl::Pose();
    unsigned long long ts_found = 0;
    if (camWorldPoseMap_ms.find(timestamp)!=camWorldPoseMap_ms.end()) {
        ts_found = timestamp;
        pose = camWorldPoseMap_ms[timestamp];
    }
    return pose;
}

sl::Pose BatchSystemHandler::findClosestLocalPoseFromTS(unsigned long long timestamp) {
    sl::Pose pose = sl::Pose();
    unsigned long long ts_found = 0;
    if (camLocalPoseMap_ms.find(timestamp)!=camLocalPoseMap_ms.end()) {
        ts_found = timestamp;
        pose = camLocalPoseMap_ms[timestamp];
    }
    return pose;
}

///
/// \brief ingestInObjectsQueue : convert a list of batched objects from SDK getObjectsBatch() to a sorted list of sl::Objects
/// \n Use this function to fill a std::deque<sl::Objects> that can be considered and used as a stream of objects with a delay.
/// \param batch_ from getObjectsBatch()
///
void BatchSystemHandler::ingestInObjectsQueue(std::vector<sl::ObjectsBatch> &batch_) {
    // If list is empty, do nothing.
    if (batch_.empty())
        return;

    // add objects in map with timestamp as a key.
    // This ensure
    std::map<uint64_t,sl::Objects> list_of_newobjects;
    for(auto &current_traj: batch_) {

        // Impossible (!!) but still better to check...
        if (current_traj.timestamps.size()!=current_traj.positions.size())
            continue;

        //For each sample, construct a objetdata and put it in the corresponding sl::Objects
        for (int j=0;j<current_traj.timestamps.size();j++) {
            sl::Timestamp ts = current_traj.timestamps.at(j);
            sl::ObjectData newObjectData;
            newObjectData.id = current_traj.id;
            newObjectData.tracking_state = current_traj.tracking_state;
            newObjectData.position = current_traj.positions.at(j);
            newObjectData.label = current_traj.label;
            newObjectData.sublabel = current_traj.sublabel;

            newObjectData.bounding_box_2d.clear();
			for (int p = 0; p < current_traj.bounding_boxes_2d.at(j).size(); p++) {
				newObjectData.bounding_box_2d.push_back(current_traj.bounding_boxes_2d.at(j).at(p));
			}

            newObjectData.bounding_box.clear();
            for (int k=0;k<current_traj.bounding_boxes.at(j).size();k++)
                newObjectData.bounding_box.push_back(current_traj.bounding_boxes.at(j).at(k));


            if (list_of_newobjects.find(ts.getMilliseconds())!=list_of_newobjects.end())
                list_of_newobjects[ts.getMilliseconds()].object_list.push_back(newObjectData);
            else
            {
                sl::Objects current_obj;
                current_obj.timestamp.setMilliseconds(ts.getMilliseconds());
                current_obj.is_new = true;
                current_obj.is_tracked = true;
                current_obj.object_list.push_back(newObjectData);
                list_of_newobjects[ts.getMilliseconds()] = current_obj;
            }
        }
    }

    // Ingest in Queue of objects that will be empty by the main loop
    // Since std::map is sorted by key, we are sure that timestamp are continous.
    for (auto &elem : list_of_newobjects)
       objects_tracked_queue.push_back(elem.second);
}

sl::Mat BatchSystemHandler::findClosestImageFromTS(unsigned long long timestamp) {
    sl::Mat image;
    unsigned long long ts_found = 0;
	if (imageMap_ms.find(timestamp) != imageMap_ms.end()) {
		ts_found = timestamp;
		image = imageMap_ms[timestamp];
	}
    return image;
}

sl::Mat BatchSystemHandler::findClosestDepthFromTS(unsigned long long timestamp) {
    sl::Mat depth;
    unsigned long long ts_found = 0;
    if (depthMap_ms.find(timestamp)!=depthMap_ms.end()) {
        ts_found = timestamp;
        depth = depthMap_ms[timestamp];
    }
    return depth;
}

File: object detection/birds eye viewer/python/batch_system_handler.py
from collections import deque
import numpy as np
import pyzed.sl as sl
import cv2

WITH_IMAGE_RETENTION = True

class BatchSystemHandler:
    '''
    The BatchSystemHandler class
    This class will transform a batch of objects (list[sl.ObjectsBatch]) from batching system into a queue/stream of sl.Objects.
    This class also handles pose of the camera in order to retrieve the pose of the camera at the object timestamp (obviously in the past).
    If WITH_IMAGE_RETENTION parameter is set to True, the images/depth/point cloud will be stored via the push() and pop() synchonized to the data.
    
    Note that image retention consumes a lot of CPU and GPU memory since we need to store them in memory so that we can output them later.
    As an example, for a latency of 2s, it consumes between 500Mb and 1Gb of memory for CPU and same for GPU. Make sure you have enought space left on memory.
    '''
    def __init__(self, data_retention_time):
        '''
        BatchSystemHandler constructor
        
        Parameters:
            data_retention_time (int): time to keep data in queue (in seconds)
        '''
        self.batch_data_retention = data_retention_time

        self.init_app_ts = sl.Timestamp()
        self.init_queue_ts = sl.Timestamp()
        self.objects_tracked_queue = deque()    # deque of sl.Objects
        self.cam_world_pose_map_ms = {}         # dict timestamp -> sl.Pose
        self.cam_local_pose_map_ms = {}         # dict timestamp -> sl.Pose
        self.image_map_ms = {}                  # dict timestamp -> image sl.Mat
        self.depth_map_ms = {}                  # dict timestamp -> depth sl.Mat

    def __del__(self):
        '''
        BatchSystemHandler destructor
        '''
        self.clear()

    def clear(self):
        '''
        clear
            Clears the remaining data in queue (free memory).
            Make sure it is called before zed is closes, otherwise you will have memory leaks
        '''
        self.objects_tracked_queue.clear()
        self.cam_world_pose_map_ms.clear()
        self.cam_local_pose_map_ms.clear()

        for key in list(self.image_map_ms.keys()):
            self.image_map_ms[key].free(sl.MEM.CPU)
            del self.image_map_ms[key]

        for key in list(self.depth_map_ms.keys()):
            self.depth_map_ms[key].free(sl.MEM.CPU)
            del self.depth_map_ms[key]

    def push(self, local_pose, world_pose, image, pc, batch):
        '''
        push
            push data in the FIFO system

        Parameters:
            local_pose (sl.Pose): current pose of the camera in camera reference frame
            world_pose (sl.Pose): current pose of the camera in world reference frame
            image (sl.Mat): image data
            pc (sl.Mat): point cloud data
            batch (list[sl.ObjectsBatch]): from ZED SDK batching system
        '''
        self.ingest_world_pose_in_map(world_pose)
        self.ingest_local_pose_in_map(local_pose)

        if WITH_IMAGE_RETENTION == True:
            self.ingest_image_in_map(world_pose.timestamp,image)
            self.ingest_depth_in_map(world_pose.timestamp,pc)
        
        self.ingest_in_objects_queue(batch)

    def pop(self, local_pose, world_pose, image, depth, objects):
        '''
        pop
            pop data from the FIFO system
        
        Parameters:
            local_pose (sl.Pose): pose of the camera in camera reference frame at objects timestamp
            world_pose (sl.Pose): pose of the camera in world reference frame at objects timestamp
            image (sl.Mat): image data at objects timestamp
            depth (sl.Mat): depth data at objects timestamp
            objects (sl.Objects): objects in the past
        '''
        objects = sl.Objects()
        local_pose = sl.Pose()
        world_pose = sl.Pose()

        if self.objects_tracked_queue:
            tracked_merged_obj = self.objects_tracked_queue[0]
            if (self.init_queue_ts.data_ns == 0):
                self.init_queue_ts = tracked_merged_obj.timestamp
        
            targetTS_ms = tracked_merged_obj.timestamp.get_milliseconds()

            local_pose = self.find_closest_local_pose_from_ts(targetTS_ms)
            world_pose = self.find_closest_world_pose_from_ts(targetTS_ms)
        
            if WITH_IMAGE_RETENTION:
                tmp_image = self.find_closest_image_from_ts(targetTS_ms)
                tmp_image.copy_to(image)
                tmp_image.free(sl.MEM.CPU)
                self.image_map_ms[targetTS_ms].free(sl.MEM.CPU)
                del self.image_map_ms[targetTS_ms]

                tmp_depth = self.find_closest_depth_from_ts(targetTS_ms)
                tmp_depth.copy_to(depth)
                tmp_depth.free(sl.MEM.CPU)
                self.depth_map_ms[targetTS_ms].free(sl.MEM.CPU)
                del self.depth_map_ms[targetTS_ms]

            objects = tracked_merged_obj
            self.objects_tracked_queue.popleft()

        return local_pose, world_pose, image, depth, objects

    def push_batch(self, batch):
        '''
        push_batch
            push data (objects data only) in the FIFO system
        
        Parameters:
            batch (list[sl.ObjectsBatch]): from ZED SDK batching system
        '''          
        self.ingest_in_objects_queue(batch)

    def pop_objects(self, objects):
        '''
        pop_objects
            pop data (objects data only) from the FIFO system
        
        Parameters:
            objects (sl.Objects): objects in the past
        '''
        if self.objects_tracked_queue:
            tracked_merged_obj = self.objects_tracked_queue[0]
            objects = tracked_merged_obj
            self.objects_tracked_queue.popleft()


    def ingest_world_pose_in_map(self, pose):
        '''
        ingest_world_pose_in_map

        Parameters:
            pose (sl.Pose): pose of the camera
        '''
        ts = pose.timestamp
        if self.init_app_ts.data_ns == 0:
            self.init_app_ts = ts

        for key in list(self.cam_world_pose_map_ms.keys()):
            if (key < (ts.get_milliseconds() - self.batch_data_retention * 1000)):
                del self.cam_world_pose_map_ms[key]
            else:
                continue
        
        self.cam_world_pose_map_ms[ts.get_milliseconds()] = pose
    
    def ingest_local_pose_in_map(self, pose):
        '''
        ingest_local_pose_in_map

        Parameters:
            pose (sl.Pose): pose of the camera
        '''
        ts = pose.timestamp
        if self.init_app_ts.data_ns == 0:
            self.init_app_ts = ts

        for key in list(self.cam_local_pose_map_ms.keys()):
            if (key < (ts.get_milliseconds() - self.batch_data_retention * 1000)):
                del self.cam_local_pose_map_ms[key]
            else:
                continue
        
        self.cam_local_pose_map_ms[ts.get_milliseconds()] = pose

    def ingest_in_objects_queue(self, batch):
        '''
        ingest_in_objects_queue: 
            Converts a list of batched objects from SDK get_objects_batch() to a sorted list of sl.Objects
            Use this function to fill a deque of sl.Objects that can be considered and used as a stream of objects with a delay
        
        Parameters:
            batch (list[sl.ObjectsBatch]): list of sl.ObjectsBatch objects obtained from calling get_objects_batch
        '''
        # If the list is empty, do nothing
        if not batch:
            return

        # Add objects in dict with timestamp as key
        list_of_new_objects = {}
        for current_traj in batch:
            # Impossible but still better to check
            if len(current_traj.timestamps) != len(current_traj.positions):
                continue
            # For each sample, construct an ObjectData and put it in the corresponding sl.Objects
            for i in range(len(current_traj.timestamps)):
                ts = current_traj.timestamps[i]
                new_object_data = sl.ObjectData()
                new_object_data.id = current_traj.id
                new_object_data.tracking_state = current_traj.tracking_state
                new_object_data.position = current_traj.positions[i]
                new_object_data.label = current_traj.label
                new_object_data.sublabel = current_traj.sublabel
                new_object_data.bounding_box_2d = current_traj.bounding_boxes_2d[i]
                new_object_data.bounding_box = current_traj.bounding_boxes[i]

                # Check if a detected object with the current timestamp already exists in the map
                if (ts.get_milliseconds() in list_of_new_objects.keys()):
                    # Append new_object_data to the object_list
                    list_of_new_objects[ts.get_milliseconds()].object_list = list_of_new_objects[ts.get_milliseconds()].object_list + [new_object_data]
                else:
                    current_obj = sl.Objects()
                    current_obj.timestamp = ts.data_ns      # sl.Objects.timestamp can be set with time in nanoseconds 
                    current_obj.is_new = True
                    current_obj.is_tracked = True 
                    # Append new_object_data to the object_list
                    current_obj.object_list = current_obj.object_list + [new_object_data]
                    list_of_new_objects[ts.get_milliseconds()] = current_obj
        
        # Ingest in Queue of objects that will be emptied by the main loop
        # Since dicts are sorted by key, we are sure that timestamps are continous.
        for value in list_of_new_objects.values():
            self.objects_tracked_queue.append(value)

    def ingest_image_in_map(self, ts, image):
        '''
        ingest_image_in_map

        Parameters:
            ts (sl.Timestamp)
            image (sl.Mat)
        '''
        self.image_map_ms[ts.get_milliseconds()] = sl.Mat()
        self.image_map_ms[ts.get_milliseconds()].clone(image)

        for key in list(self.image_map_ms.keys()):
            if key < (ts.get_milliseconds() - self.batch_data_retention * 1000 * 2):
                self.image_map_ms[key].free(sl.MEM.CPU)
                del self.image_map_ms[key]
            else:
                continue

    def ingest_depth_in_map(self, ts, depth):
        '''
        ingest_depth_in_map

        Parameters:
            ts (sl.Timestamp)
            depth (sl.Mat)
        '''
        self.depth_map_ms[ts.get_milliseconds()] = sl.Mat()
        self.depth_map_ms[ts.get_milliseconds()].clone(depth)

        for key in list(self.depth_map_ms.keys()):
            if key < (ts.get_milliseconds() - self.batch_data_retention * 1000 * 2):
                self.depth_map_ms[key].free(sl.MEM.CPU)
                del self.depth_map_ms[key]
            else:
                continue

    def find_closest_local_pose_from_ts(self, timestamp):
        '''
        find_closest_local_pose_from_ts
            Find the sl.Pose (in camera reference frame) that matches the given timestamp
        
        Parameters:
            timestamp (int): timestamp in ms (or at least in the same unit as self.cam_local_pose_map_ms)
        Return:
            The matching sl.Pose
        '''
        pose = sl.Pose()
        if timestamp in self.cam_local_pose_map_ms.keys():
            pose = self.cam_local_pose_map_ms[timestamp]
        return pose

    def find_closest_world_pose_from_ts(self, timestamp):
        '''
        find_closest_world_pose_from_ts
            Find the sl.Pose (in world reference frame) that matches the given timestamp
        
        Parameters:
            timestamp (int): timestamp in ms (or at least in the same unit as self.cam_world_pose_map_ms)
        Return:
            The matching sl.Pose
        '''
        pose = sl.Pose()
        if timestamp in self.cam_world_pose_map_ms.keys():
            pose = self.cam_world_pose_map_ms[timestamp]
        return pose

    def find_closest_image_from_ts(self, timestamp):
        '''
        find_closest_image_from_ts
            Find the sl.Mat that matches the given timestamp
        
        Parameters:
            timestamp (int): timestamp in ms (or at least in the same unit as self.image_map_ms)
        Return:
            The matching sl.Mat
        '''
        image = sl.Mat()
        if timestamp in self.image_map_ms.keys():
            image = self.image_map_ms[timestamp]
        return image

    def find_closest_depth_from_ts(self, timestamp):
        '''
        find_closest_depth_from_ts
            Find the sl.Mat that matches the given timestamp
        
        Parameters:
            timestamp (int): timestamp in ms (or at least in the same unit as self.depth_map_ms)
        Return:
            The matching sl.Mat
        '''
        depth = sl.Mat()
        if timestamp in self.depth_map_ms.keys():
            depth = self.depth_map_ms[timestamp]
        return depth


    

    